### ISSUE TITLE
High TR (header retrieval time) using HTTP/2 and HAproxy 2.4

### ISSUE THREAD

**Issue by bartbutler** [haproxy/haproxy#1395]
### Detailed Description of the Problem

We have tried to update our HAproxy installations to 2.4. HTTP/1.1 is OK, but when we go from 2.2 to 2.4 we get a large jump in 503s to clients. It's not a large fraction relative to total requests and seems random (the requests don't have anything in common and retries work) but it's significant relative to 2.2 levels. The main characteristic of the additional 503s seem to be relatively large TR (hundreds of milliseconds up to 10s or more, approximately equal to Ta) followed by client disconnection (CC--).

Example log timing information (CLF):

```
TR     Tw Tc Tr  Ta       tsc
391    0   -1  -1  391    CC--
492   0   -1  -1  494    CC--
159    0   -1  -1  162    CC--
5409 0   -1  -1  5409 CC--
```

A significant number of requests, not necessarily the 503s, are more than 30s in TR, so it seems like the increase in 503s is just due to an overall drastic increase in TR across the board.

### Expected Behavior

TR to be very small, 1ms usually

### Steps to Reproduce the Behavior

Upgrade to 2.4 from 2.2 with HTTP/2 on.

### What is your configuration?

```haproxy
global:
    log         127.0.0.1 local1 notice alert
    # increase length (default 1024) for logstash issue with truncated logs
    log         127.0.0.1 len 2048 local2 info alert

    chroot      /var/lib/haproxy
    pidfile     /var/run/haproxy.pid
    maxconn     200000
    user        haproxy
    group       haproxy
    daemon

    # ssl related
    maxsslconn  500000
    ssl-engine rdrand  # intel engine available in openssl
    ssl-mode-async
    tune.ssl.default-dh-param 2048
    ## ciphers list from https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
    ssl-default-bind-options no-tls-tickets ssl-min-ver TLSv1.2
    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256

    ## turn on stats unix sockets, one per process
    stats socket /var/lib/haproxy/stats mode 0660 user haproxy group haproxy level admin

    tune.maxrewrite 1280

defaults
    mode                    http
    log                     global
    no option               dontlognull
    option                  http-server-close
    option                  redispatch 1
    option                  allbackups

    retries                 3
    timeout http-request    180s
    timeout queue           180s
    timeout connect         3500
    timeout client          180s
    timeout server          180s
    timeout client-fin      1s
    timeout server-fin      1s
    timeout http-keep-alive 180s
    timeout check           5s
    maxconn                 200000
    option                  abortonclose
    option                  tcp-smart-connect
    hash-balance-factor     125
    hash-type               consistent
    ## This option prevents 408 errors in Chrome/IE due to pre-connects
    ## See: http://blog.haproxy.com/2014/05/26/haproxy-and-http-errors-408-in-chrome/
    errorfile 408 /dev/null

    ## default settings for server checks
    default-server ca-file check check-ssl ssl ssl-min-ver TLSv1.2 verify required
    default-server resolve-prefer ipv4

frontend webapp-live-https
    ## anycast
    bind        123.123.123.123:443  tfo alpn h2,http/1.1 ssl crt /etc/pki/haproxy/LE.ev.example.com.chain.pem
    bind        123.123.123.123:4005 tfo ssl crt /etc/pki/haproxy/LE.ev.example.com.chain.pem accept-proxy
    bind        123.123.123.123:4305 tfo ssl crt /etc/pki/haproxy/LE.ev.example.com.chain.pem accept-proxy

    mode        http
    ## CLF format extended with TLS protocol and ciphers used information
    log-format  "%{+Q}o %{-Q}ci - - [%trg] %r %ST %B \"\" \"\" %cp %ms %ft %b %s %TR %Tw %Tc %Tr %Ta %tsc %ac %fc %bc %sc %rc %sq %bq %CC %CS %hrl %hsl %sslv %sslc %[ssl_fc_session_id,base64] %[ssl_fc_sni]"

    tcp-request inspect-delay 5s # leave up to this amount if time to get the request path
    
    # various acls and header manipulation

    default_backend webapp-live-https-servers

backend webapp-live-https-servers
    mode        http
    balance     source
    option      httpchk HEAD /httpchk HTTP/1.1
    http-check send hdr Host example.com
    http-check disable-on-404

    default-server sni str(example.com) check-sni example.com  maxconn 200 maxqueue 1
    server  webapp-live_016    web-016.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_020    web-020.data.:443   weight "${GV2_M_WEIGHT}"
    server  webapp-live_021    web-021.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_024    web-024.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_027    web-027.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_028    web-028.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_029    web-029.data.:443   weight "${GV2_M_WEIGHT}"
    server  webapp-live_031    web-031.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_032    web-032.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_033    web-033.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_034    web-034.data.:443   weight "${GV2_M_WEIGHT}"
    server  webapp-live_037    web-037.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_038    web-038.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_039    web-039.data.:443   weight "${ZH5_L_WEIGHT}" maxconn 400
    server  webapp-live_040    web-040.data.:443   weight "${ZH5_M_WEIGHT}"
    server  webapp-live_041    web-041.data.:443   weight "${ZH5_L_WEIGHT}" maxconn 400
    server  webapp-live_042    web-042.data.:443   weight "${ZH5_L_WEIGHT}" maxconn 400
    server  webapp-live_043    web-043.data.:443   weight "${ZH5_L_WEIGHT}" maxconn 400
    server  webapp-live_044    web-044.data.:443   weight "${ZH5_L_WEIGHT}" maxconn 400
    server  webapp-live_045    web-045.data.:443   weight "${FR3_L_WEIGHT}" maxconn 400
    server  webapp-live_046    web-046.data.:443   weight "${FR3_L_WEIGHT}" maxconn 400
    server  webapp-live_047    web-047.data.:443   weight "${FR3_M_WEIGHT}"
    server  webapp-live_048    web-048.data.:443   weight "${FR3_M_WEIGHT}"
    server  webapp-live_049    web-049.data.:443   weight "${FR3_M_WEIGHT}"
    server  webapp-live_050    web-050.data.:443   weight "${FR3_M_WEIGHT}"
    server  webapp-live_051    web-051.data.:443   weight "${FR3_M_WEIGHT}"
    server  webapp-live_052    web-052.data.:443   weight "${FR3_M_WEIGHT}"
    server  webapp-live_053    web-053.data.:443   weight "${FR3_L_WEIGHT}" maxconn 400
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.4-acb1d0b 2021/09/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.4.html
Running on: Linux 5.10.16-1.el7.elrepo.x86_64 #1 SMP Thu Feb 11 17:42:15 EST 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_PCRE=1 USE_PTHREAD_PSHARED=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_TFO=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD +PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=64).
Built with OpenSSL version : OpenSSL 1.1.1l  24 Aug 2021
Running on OpenSSL version : OpenSSL 1.1.1l  24 Aug 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.4.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with zlib version : 1.2.7
Running on zlib version : 1.2.7
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 4.8.5 20150623 (Red Hat 4.8.5-44)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by wtarreau** [comment]
Note, I've got a separate report in private that matches this, which is why there's no response for now. We're still trying to figure what's happening. The ticket will be updated once there are any news.

**Comment by bartbutler** [comment]
Thanks @wtarreau !

**Comment by wtarreau** [comment]
So the one I was thinking about it this one: https://github.com/haproxy/haproxy/commit/cb59e0

However @capflam thinks that theoretically it shouldn't cause any problem in 2.4. The bug are so similar-looking that I think it would be worth double-checking, as often there are other corner cases that are overlooked and let the code express itself in a tighter situation. Above you're seeing these 5s timeouts and that's the timeout of your TCP rules... That sounds extremely familiar :-)

You seem to be compiling your executables yourself, so I'm assuming you're at ease with applying a patch. Otherwise just wait for 2.4.5 in a day or two.

**Comment by bartbutler** [comment]
Thanks @wtarreau. It does sound familiar but I would note that we also see much longer TR (the logs above are just an example of the distribution shifting), so I'm not sure that there is any sort of obvious peak around 5s. I can try to get a distribution next time we test. We'll happily try out 2.4.5 regardless and see if it helps.

**Comment by wtarreau** [comment]
OK, do not worry for now regarding the distribution as I don't know either if other elements may cause artefacts (e.g. browser preconnect). The issue that used to happen was with H2 clients that would not use ALPN and would thus reach the H1 mux then upgrade to H2. The tcp-request rules might have started there before the mux decides to kill the stream and upgrade. That's a totally messy case. Normally 2.4 is waayyyyy cleaner than previous versions on this upgrade, but as I said, corner cases can never be totally ruled out.

Indeed, testing 2.4.5 will simplify all of this troubleshooting. If it continues we might have to figure another cause but we won't have to think about this one anymore :-)

**Comment by capflam** [comment]
FYI, the 2.4.7 was released (2.4.5 and 2.4.6 are affected by regressions and must not be deployed).

**Comment by bartbutler** [comment]
It took us a few days to test but unfortunately we do not see any improvement in this issue with 2.4.7.

**Comment by wtarreau** [comment]
OK, thanks for testing. There are 3 things that still strike me in your config:
- `ssl-engine rdrand`: this one is wrong, it stems from a misunderstanding of engines that ended up in a presentation a long time ago and that continues to pop up from time to time. We're not aware of particular issues it could cause but it's hard to tell ; please remove it.
- `ssl-async` : if you don't use a real engine, this one is useless. But as it involves completely separate paths at the transport layer, it could definitely be possible that we'd have an issue there on outgoing connections, for example with connections not being able to migrate between threads (just an idea). Please remove it.
- `maxqueue 1`: it can cause 503s if all your servers are in use, but I'll assume that this is not the case. However it could be possible that we slightly broke something related to it with the queues, so could you please try to temporarily disable it to verify if it's in any way related ?

Also, do you have any backend with clear-text servers to compare if using SSL plays a role in these 503 at all ?

**Comment by bartbutler** [comment]
Hi @wtarreau , thanks for the advice. We'll try all of those (the maxqueue 1 was our attempt to entirely disable per-user stickiness if the server in question was full). We don't have any cleartext backends, no.

**Comment by TimWolla** [comment]
We don't actively monitor this, but seeing the title of the recent email notification reminded me that I've seen this issue (high TR) as well. I am not seeing a 503. However it's annoying, because it makes the logs a bit useless.

I can reproduce this as follows:

1. Load the website in Google Chrome.
2. Open Chrome's dev tools.
3. Reload HAProxy to tear down all keep-alive H2 connections.
4. Right click the reload button.
5. Empty Cache and Hard Reload
6. Chrome establishes a new H2 connection with regular timings.
7. Wait a few seconds.
8. Right click the reload button.
9. Empty Cache and Hard Reload
10. HAProxy shows an incorrect value for TR:

These are the (redacted) logs by grepping for my own IP address:

```
Oct 25 11:57:56 *snip* haproxy[27728]: *snip*:49652 [25/Oct/2021:11:57:56.502] https~ *snip* 414/0/0/2/415 200 15286 - - ---- 14/13/0/0/0 0/0 {www.example.com|Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36} "GET https://www.example.com/*snip* HTTP/2.0" S1-61769BC4002E
Oct 25 11:57:56 *snip* haproxy[27728]: *snip*:49668 [25/Oct/2021:11:57:56.937] https~ *snip* 2/0/0/1/25 200 428 - - ---- 15/14/0/0/0 0/0 {www.example.com|Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36} "GET https://www.example.com/*snip* HTTP/2.0" S1-61769BC4002F
Oct 25 11:58:09 *snip* haproxy[27728]: *snip*:49652 [25/Oct/2021:11:57:56.918] https~ *snip* 12238/0/0/41/12278 200 18561 - - ---- 25/25/0/0/0 0/0 {www.example.com|Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36} "GET https://www.example.com/*snip* HTTP/2.0" S1-61769BC4008A
```

```
HAProxy version 2.4.7-1~bpo10+1 2021/10/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.7.html
Running on: Linux 4.19.0-17-amd64 #1 SMP Debian 4.19.194-3 (2021-07-18) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-2.4.7=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=12).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by kotipsy** [comment]
maxqueue was introduced after testing the new version and indeed 503 with 2.4.x aren't related to potential backend load in this case.
About ssl , we don't have traffic in clear-text with backend
Then we've removed the 2 params `ssl-async` and `ssl-engine rdrand` + retry the upgrade and same result. 
Additional comment  on this issue is more of  those 503 occur on css/png/js compared to almost 0 for API request type. 
Same request from different users (http version, tls version, ciphers) can return 200 or 503
Could http2 (frontend) to http1.1 (backend )could play a role here ?

**Comment by kotipsy** [comment]
Hello,
Some updates on this issue.
  My last question didn't make sense since the requests failed a way before being send to the backend
I saw you've released the 2.4.8. I'll replay the upgrade with this version and see if it helps.
In between, I've played with h2 header trying to set higher max.
```
    tune.bufsize 32768
    tune.ssl.cachesize 500000
    tune.h2.header-table-size 65536
```
(didn't play with stream size to `0` as too risky on prod)

Then I tried to trace on both version (2.2.17 and 2.4.7) using following trace settings
```
trace_type="h2"
trace_event="+any"
trace_verbosity="advanced"
trace_level="state"
trace_sink="stdout"
```
I see I/O h2 proto errors on both but nothing higher on the latest version compared to the running one.
What I do know is:
- Sampling 1 sec of tracing I could see in the Haproxy log 3 clients getting 503. Those clients, I can't retrieve them on the trace output (I played the trace at different time during few secs and found the same issue).
- I've compared other patterns (beside most of the 503 are js/css/png/etc compared to API calls) and found out that most of the requests use TLS1.3.

Sample of traffic: 48%.  VS TLS1.2 52%
Percentage of 503: 98% TLS1.3. VS  2% TLS1.2

Any other ideas I could check come up in your mind ?

**Comment by wtarreau** [comment]
Interesting, maybe this could have something to do with something around 0-rtt, but I'm not seeing any in your config. I don't have any smart idea of anything to test on the subject for now however, we'll need to think about this a bit more.

Hmmm could you possibly try with `no-tlsv13` on the bind line just to see if the problem disappears, significantly drops, or doesn't change ?

**Comment by kotipsy** [comment]
Ok, I'll try it this week-end with the latest version

**Comment by kotipsy** [comment]
Hello,
I've tried what we said by disabling TLS 1.3 in bind after deploying 2.4.8.
I confirm the number of 503 increased significantly for TLS 1.2 but less compare to TLS 1.3.
I've tried to compile with openssl 3.0 to see if it helps but Haproxy code isn't yet compatible.
I only replayed once, will replay it on test lab a bit long for more inputs.

**Comment by EmericBr** [comment]
Just for note: ssl-mode-async is useless on engine rdrand, since this engine doesn't use a device and was not blocking. ssl-mode-async was designed to work using intel quick assist engine.

**Comment by lseelenbinder** [comment]
Hi all, @wtarreau,

Just wanted to confirm I'm seeing high `TR` values in 2.4 & 2.5 with HTTP2. However, these are spurious, because it's counting the time between two requests if the connection remained open.

For example, if I request 10 files at T0, wait 10 seconds, and request 10 more at T1. The first request of T1 will show 10+ seconds for the TR value. The rest of the values are correct.

Hope this helps.

**Comment by wtarreau** [comment]
Hi Luke,

that's very useful, so it means that for some reason the idle time is incorporated in the response time calculation. I see how that can happen, I need to have a closer look at this.

Thanks!

**Comment by bartbutler** [comment]
That could very well explain our 503 spikes, as I imagine timeouts are also based on the response time calculation. Thanks!

**Comment by wtarreau** [comment]
I was about to say that it ought to be unrelated, but am now rolling that back. In fact if the request arrives with an apparent age that is higher than the request timeout, it might possibly be aborted during the attempt to connect to the server (or to be queued there). So yes, this remains an interesting track to study!

**Comment by lseelenbinder** [comment]
If it helps any, the frontends are H2 and the backends are H1. Not sure if the translation could introduce the issue. :shrug:

**Comment by bartbutler** [comment]
For us as well.

**Comment by wtarreau** [comment]
OK thank you guys for your help and sorry for the delay. Here's a fix for this. It works fine for me and matches exactly what's mentioned in the doc, i.e. that idle starts to count immediately after the previous request. The alternative would be to count when there are no streams at all but it would violate the doc and be less useful (here we're seeing js/css being fetched a few tens of ms after the start of the retrieval of an HTML file for example). Please let me know if that's fine for you, and if so I'll merge it. If you both prefer the 100% idle time only (i.e. no stream) then this must be discussed with other users first.
[0001-BUG-MINOR-mux-h2-update-the-session-s-idle-delay-bef.patch.txt](https://github.com/haproxy/haproxy/files/8001205/0001-BUG-MINOR-mux-h2-update-the-session-s-idle-delay-bef.patch.txt)

**Comment by bartbutler** [comment]
Sounds reasonable to me. We can try it out and see if the downstream issues we had go away as well.

**Comment by lseelenbinder** [comment]
As long as `TR` doesn't include the idle time (and thus get counted in total response time of `Ta`), any solution works for us. Thanks @wtarreau!

**Comment by wtarreau** [comment]
The fix was now merged to 2.6, 2.5 and 2.4. You have it in the fresh 2.4.13. I'm closing the issue now, thanks for your participation guys!

**Comment by lseelenbinder** [comment]
Deployed today and confirmed it works @wtarreau. Thank you!
